File: HP_Noise_Figure_Meter_8970B/Noise_Figure_Meter_8970B.py
# Any problem, contact jxli@bao.ac.cn
# Last modified 2018/10/26
import argparse
import time
from sys import stdout as out

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Controller for HP Noise Figure Meter 8970B')
    parser.add_argument('-e', dest='enr', nargs='?', const='agilent', choices=('agilent', 'other'), help='Input ENR parameters; Default ENR: Agilent 346A, "other" is not supported')
    parser.add_argument('-c', dest='calibrate', metavar=('StartFrequency', 'StopFrequency', 'StepSize'), nargs=3, type=int, help='Calibrate with StartFrequency StopFrequency StepSize; Unit: MHz; Default Smoothing: 5, use -s to assign smoothing times.')
    parser.add_argument('-s', dest='smoothing', metavar='Smoothing', choices=range(10), type=int, default=5, help='Set 2**Smoothing; Default: 5')
    parser.add_argument('-m', dest='measure', type=int, metavar=('StartFrequency', 'StopFrequency', 'StepSize'), nargs=3, help='Automatically measure with StartFrequency StopFrequency StepSize and save result to file; Unit: MHz; Default Filename: NF_data_$CurrentDateTime$.txt, use -d to assign datafile name.')
    parser.add_argument('-d', dest='datafile', metavar='DatafileName', default=time.strftime('NF_data_%Y%m%d%H%M%S.txt'), help='Data file name; Default: NF_data_$CurrentDateTime$.txt')
    parser.add_argument('-t', dest='time', metavar='Time', default=6, type=int, help='Set waiting time for smoothing process before reading data; Unit: second; Default: 6; Larger smoothing, longer time.')
    parser.add_argument('-f', dest='frequency', metavar='Frequency', type=int, help='Manually measure at assigned Frequency; Unit: MHz')
    parser.add_argument('-p', dest='plot', metavar='DatafileName', nargs='?', const='NF_data_', help='Plot noise figure data with DatafileName; Default: NF_data_$LatestDateTime$.txt')
    options = parser.parse_args()
    #print options
    
    if options.plot != None:
        if options.plot == 'NF_data_':
            import glob
            txtfiles = glob.glob('NF_data_??????????????.txt')
            if txtfiles == []:
                print 'Error: Can not find data file of format NF_data_$LatestDateTime$.txt .'
                exit()
            txtfile = sorted(txtfiles)[-1]
        else:
            txtfile = options.plot
        print('Loading {0} ...'.format(txtfile))
        import numpy as np
        data = np.loadtxt(txtfile, delimiter = ',')
        datetime_str = txtfile[:-4]
        if datetime_str.startswith('NF_data_'):
            datetime_str = datetime_str[8:]
        import matplotlib.pyplot as plt
        plt.figure(figsize = (12, 8))
        plt.title('Gain')
        plt.plot(data[:, 0] / 1e6, data[:, 1])
        plt.grid()
        plt.xlabel('Frequency (MHz)')
        plt.ylabel('Gain (dB)')
        gain_graph = 'Gain_{0}.png'.format(datetime_str)
        plt.savefig(gain_graph)
        print(gain_graph)

        plt.figure(figsize = (12, 8))
        plt.title('Noise Figure')
        plt.plot(data[:, 0] / 1e6, data[:, 2])
        plt.grid()
        plt.xlabel('Frequency (MHz)')
        plt.ylabel('Noise Figure (dB)')
        nf_graph = 'NoiseFigure_{0}.png'.format(datetime_str)
        plt.savefig(nf_graph)
        print(nf_graph)
        print('Over.')
        exit()
   
    if options.calibrate == None and options.enr == None and options.frequency == None and options.measure == None:
        parser.print_help()
        exit()

    import pyvisa
    rm = pyvisa.ResourceManager()
    src = rm.list_resources()
    if src == ():
        print('Error: cannot connect to instrument. Check your GPIB connection.')
        exit()
    else:
        # By now, only one GPIB device is supported. If two or more GPIB devices are used, you should manually set it here by:
        # dev = src[index] # where index is the one you wanted. Then comment the for loop below.
        for dev in src:
            if 'GPIB' in dev:
                break
            else:
                print('Error: cannot find GPIB device. Check your GPIB connection.')
                exit()

    if options.enr != None:
        if options.enr.lower() == 'agilent':
            # ENR for Agilent 346A 10MHz-18GHz Noise Source.    
            ENR_str = 'NR10MZ5.47EN100MZ5.56EN1000MZ5.37EN2000MZ5.41EN3000MZ5.42EN4000MZ5.40EN5000MZ5.32EN6000MZ5.39EN7000MZ5.46EN' + \
            '8000MZ5.39EN9000MZ5.53EN10000MZ5.62EN11000MZ5.51EN12000MZ5.64EN13000MZ5.68EN14000MZ5.77EN15000MZ5.79EN16000MZ5.91EN17000MZ5.84EN18000MZ5.40EN'
        elif options.enr.lower() == 'other':
            print('Error: ENR for "other" is unknown.')
            ENR_str = 'NRxx.xxMZx.xxEN'
            exit()
        else:
            print('Error: unsupported ENR.')
            exit()
        inst = rm.open_resource(dev)
        status = inst.write(ENR_str)
        time.sleep(1)
        # Special function sets (Fre, Gain, NF) output.
        status = inst.write('SPH1')

        print 'Set ENR succeeded.'
        try:
            inst.close()
        except:
            print('Warn: disconnect instrument failed.')
    elif options.calibrate != None:
        inst = rm.open_resource(dev)
        status = inst.write('M2FA{0[0]}MZFB{0[1]}MZSS{0[2]}MZF{1}CA'.format(options.calibrate, options.smoothing))
        print('Set calibration succeeded.')
        print('Wait the instrument to finish calibration before carrying on.')
        try:
            inst.close()
        except:
            print('Warn: disconnect instrument failed.')
    elif options.frequency != None:
        inst = rm.open_resource(dev)
        status = inst.write('FR{0}MZEN'.format(options.frequency))
        print('Set frequency succeeded.')
        time.sleep(options.time)
        data = inst.read().strip()
        data_fre, data_gain, data_nf = data.split(',')
        data_fre = float(data_fre)
        try:
            data_gain= float(data_gain)
            data_nf  = float(data_nf)
        except:
            print('Gain = Error dB\tNF = Error dB')
        else:
            print('Gain = {0:>6.3f}dB\tNF = {1:>6.3f}dB'.format(data_gain, data_nf))
        try:
            inst.close()
        except:
            print('Warn: disconnect instrument failed.')
    elif options.measure != None:
        inst = rm.open_resource(dev)
        # Confirm Corrected NF and Gain.
        status = inst.write('M2EN') 
        print('Start measuring ...') 
        time.sleep(1)      
        fstart, fstop, fstep = options.measure
        with open(options.datafile, 'w') as df:
            for fre in xrange(fstart, fstop+1, fstep):
                out.write('Freq = {0:>4d}MHz\t'.format(fre))
                out.flush()
                status = inst.write('FR{0}MZM2'.format(fre))
                time.sleep(options.time)
                data = inst.read().strip()
                df.write(data + '\n')
                data_fre, data_gain, data_nf = data.split(',')
                data_fre = float(data_fre)
                try:
                    data_gain= float(data_gain)
                    data_nf  = float(data_nf)
                except:
                    print('Gain = Error dB\tNF = Error dB')
                else:
                    print('Gain = {0:>6.3f}dB\tNF = {1:>6.3f}dB'.format(data_gain, data_nf))
        print('\nOver. Data has been saved to {0}'.format(options.datafile))
        try:
            inst.close()
        except:
            print('Warn: disconnect instrument failed.')
            exit()
    else:
        exit()

File: HP_Noise_Figure_Meter_8970B/Noise_Figure_Meter_8970B.m
% Initial parameters.
f_start = 1605; % MHz
f_stop = 1705; % MHz
f_step = 5; % MHz
smoothing = 5; % 2^smoothing where smoothing is in [0, 1, ...,9]
% DUT_name = 'Amplifier-ZX60-33LN-S-Step10MHz';
% DUT_name = 'test_inst';
% DUT_name = 'NF-ZX60-33LN-S-Step10-10to4000_once';
DUT_name = '54LNAold1-3';
datafile_name = ['D:\\www\', DUT_name, '.txt'];
% cmd = 'poweron';
% cmd = 'calibration';
% cmd = 'work';
cmd = 'plot';


% smoothing:6->6.5s.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% dev = gpib('ni', 0, 8);
% dev = gpib('agilent', 0, 19);
% set(dev, 'InputBufferSize', 2500);
% fopen(dev);
% function dev = opendev(vendor, board_index, address)
% dev = gpib(vendor, board_index, address);
% set(dev, 'InputBufferSize', 2500);
% devid = fopen(dev);

if strcmp(cmd,'poweron')
    % Enter ENR table.
    % ENR for Agilent 346A 10MHz-18GHz Noise Source.    
    fprintf(dev, ['NR10MZ5.47EN100MZ5.56EN1000MZ5.37EN2000MZ5.41EN',...
        '3000MZ5.42EN4000MZ5.40EN5000MZ5.32EN6000MZ5.39EN7000MZ5.46EN',...
        '8000MZ5.39EN9000MZ5.53EN10000MZ5.62EN11000MZ5.51EN',...
        '12000MZ5.64EN13000MZ5.68EN14000MZ5.77EN15000MZ5.79EN',...
        '16000MZ5.91EN17000MZ5.84EN18000MZ5.40EN']);
    pause(1);
    % Special function sets (Fre, Gain, NF) output.
    fprintf(dev, 'SPH1');
elseif strcmp(cmd,'calibration')
    % Calibration.
    cali_cmd = ['M2FA',num2str(f_start),'MZFB',num2str(f_stop),'MZSS',num2str(f_step),'MZF',num2str(smoothing),'CA']
    fprintf(dev,cali_cmd);
elseif strcmp(cmd,'work')
    % Confirm Corrected NF and Gain.
    fprintf(dev, 'M2EN');
    pause(1);
    fid = fopen(datafile_name, 'at');
    % Get data.
%     fprintf(fid, '###########################################\n');
    for i=f_start:f_step:f_stop
        work_cmd = ['FR',num2str(i),'MZM2'];
        fprintf(dev,work_cmd);
        pause(6);
        fprintf(fid, fscanf(dev));
    end
    fclose(fid);
    NFdata = importdata(datafile_name, ',', 1);
    fig_gain = figure();
    plot(NFdata.data(:,1)./1e6, NFdata.data(:,2));
    grid on;
    xlabel('Frequency(MHz)');
    ylabel('Gain(dB)');
    title(['Gain of ', DUT_name]);
    print(fig_gain,'-dpng',['D:\\www\Gain_figure_',DUT_name,'.png']);
    fig_nf =  figure();
    plot(NFdata.data(:,1)./1e6,NFdata.data(:,3));
    grid on;
    xlabel('Frequency(MHz)');
    ylabel('Noise Figure(dB)');
    title(['Noise Figure of ', DUT_name]);
    print(fig_nf,'-dpng',['D:\\www\NF_figure_',DUT_name,'.png']);
elseif strcmp(cmd, 'plot')
    NFdata = importdata(datafile_name, ',', 1);
    fig_gain = figure();
    plot(NFdata.data(:,1)./1e6, NFdata.data(:,2));
    grid on;
%     ylim([5, 25]);
    xlabel('Frequency(MHz)');
    ylabel('Gain(dB)');
    title(['Gain of ', DUT_name]);
    print(fig_gain,'-dpng',['D:\\www\Gain_figure_',DUT_name,'.png']);
    fig_nf =  figure();
    plot(NFdata.data(:,1)./1e6, NFdata.data(:,3));
    grid on;
    ylim([0.0, 3]);
    xlabel('Frequency(MHz)');
    ylabel('Noise Figure(dB)');
    title(['Noise Figure of ', DUT_name]);
    print(fig_nf,'-dpng',['D:\\www\NF_figure_',DUT_name,'.png']);
end
% fclose(dev);
% delete(dev);
